extensions/amp-position-observer/0.1/amp-position-observer.js

```diff
@@ -0,0 +1,378 @@
+   * @private
+   */
+  updateVisibility_(positionRect, adjustedViewportRect, relativePos) {
+    // Fully inside margin-adjusted viewport.
+    if (relativePos == RelativePositions.INSIDE) {
+      this.isVisible_ = true;
+      return;
+    }
+
+    const ratioToUse = relativePos == RelativePositions.TOP ?
+        this.topRatio_ : this.bottomRatio_;
+
+    const offset = positionRect.height * ratioToUse;
+    if (relativePos == RelativePositions.BOTTOM) {
+      this.isVisible_ =
+        positionRect.top <= (adjustedViewportRect.bottom - offset);
+      return;
+    } else {
+      this.isVisible_ =
+        positionRect.bottom >= (adjustedViewportRect.top + offset);
+      return;
+    }
+  }
+
+  /**
+   * Calculates the current scroll progress as a percentage.
+   * Scroll progress is a decimal between 0-1 and shows progress between
+   * enter and exit, considering ratio and margins.
+   * When a scene becomes visible (enters based on ratio and margins), from
+   * bottom, progress is 0 as it moves toward the top, progress increases until
+   * it becomes exists with 1 from the top.
+   *
+   * Entering from the top gives the reverse values, 1 at enter, 0 at exit.
+   *
+   * @private
+   */
+  updateScrollProgress_(positionRect, adjustedViewportRect) {
+    const totalDurationOffset = (positionRect.height * this.bottomRatio_) +
+        (positionRect.height * this.topRatio_);
+
+    const totalDuration = adjustedViewportRect.height +
+        positionRect.height - totalDurationOffset;
+
+    const topOffset = Math.abs(
+        positionRect.top - this.resolvedTopMargin_ -
+        (adjustedViewportRect.height -
+            (positionRect.height * this.bottomRatio_)
+        )
+      );
+
+    this.scrollProgress_ = topOffset / totalDuration;
+  }
+
+  /**
+   * @private
+   */
+  parseAttributes_() {
+    // Ratio is either "<top-bottom:{0,1}>" or "<top:{0,1}> <bottom:{0,1}>"
+    // e.g, "0.5 1": use 50% visibility at top and 100% at the bottom of viewport.
+    const ratios = this.element.getAttribute('intersection-ratios');
+    if (ratios) {
+      const topBottom = ratios.trim().split(' ');
+      this.topRatio_ = this.validateAndResolveRatio_(topBottom[0]);
+      this.bottomRatio_ = this.topRatio_;
+      if (topBottom[1]) {
+        this.bottomRatio_ = this.validateAndResolveRatio_(topBottom[1]);
+      }
+    }
+
+    // Margin is either "<top-bottom:{px,vh}>" or "<top:{px,vh}> <bottom:{px,vh}>"
+    // e.g, "100px 10vh": exclude 100px from top and 10vh from bottom of viewport.
+    const margins = this.element.getAttribute('exclusion-margins');
+    if (margins) {
+      const topBottom = margins.trim().split(' ');
+      this.topMarginExpr_ = topBottom[0];
+      this.bottomMarginExpr_ = this.topMarginExpr_;
+      if (topBottom[1]) {
+        this.bottomMarginExpr_ = topBottom[1];
+      }
+    }
+
+    this.targetSelector_ = this.element.getAttribute('target-selector');
```

Comment:
FWIW, I think the `target="id"` form is better for this use case. Selector kind of implies behavior for multiple targets. This also clarifies typical confusion around the direction of DOM search, e.g. it's not a `closest` thing, etc. Asking folk to add an ID for this use case is very reasonable. Especially since it's optional anyway.